packages/flutter_tools/lib/src/commands/build_ios.dart

```diff
@@ -130,6 +130,74 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
     return super.validateCommand();
   }
 
+  // Parses Contents.json into a map, with the key to be the combination of (idiom, size, scale), and value to be the icon image file name.
+  Map<String, String> _parseIconContentsJson(String contentsJsonDirName) {
+    final Directory contentsJsonDirectory = globals.fs.directory(contentsJsonDirName);
+    if (!contentsJsonDirectory.existsSync()) {
+      return <String, String>{};
+    }
+    final File contentsJsonFile = contentsJsonDirectory.childFile('Contents.json');
+    final Map<String, dynamic> content = json.decode(contentsJsonFile.readAsStringSync()) as Map<String, dynamic>;
+    final List<dynamic> images = content['images'] as List<dynamic>? ?? <dynamic>[];
+
+    final Map<String, String> iconInfo = <String, String>{};
+
+    for (final dynamic image in images) {
+      final Map<String, dynamic> imageMap = image as Map<String, dynamic>;
+      final String? idiom = imageMap['idiom'] as String?;
+      final String? size = imageMap['size'] as String?;
+      final String? scale = imageMap['scale'] as String?;
+      final String? fileName = imageMap['filename'] as String?;
+
+      if (size != null && idiom != null && scale != null && fileName != null) {
+        iconInfo['$idiom $size $scale'] = fileName;
+      }
+    }
+
+    return iconInfo;
+  }
+
+  Future<void> _validateIconsAfterArchive() async {
+    final BuildableIOSApp app = await buildableIOSApp;
+    final String templateIconImageDirName = await app.templateAppIconDirNameForImages;
+
+    final Map<String, String> templateIconMap = _parseIconContentsJson(app.templateAppIconDirNameForContentsJson);
+    final Map<String, String> projectIconMap = _parseIconContentsJson(app.projectAppIconDirName);
+    // find project icons that are in conflict with template icons
+    final List<MapEntry<String, String>> conflictIcons = projectIconMap.entries
+        .where((MapEntry<String, String> entry) {
+      final String projectIconFileName = entry.value;
+      final String? templateIconFileName = templateIconMap[entry.key];
+      if (templateIconFileName == null) {
+        return false;
+      }
+
+      final File projectIconFile = globals.fs.file(globals.fs.path.join(app.projectAppIconDirName, projectIconFileName));
+      final File templateIconFile = globals.fs.file(globals.fs.path.join(templateIconImageDirName, templateIconFileName));
+      return projectIconFile.existsSync()
+          && templateIconFile.existsSync()
+          && Object.hashAll(projectIconFile.readAsBytesSync()) == Object.hashAll(templateIconFile.readAsBytesSync());
```

Comment:
@christopherfujino is that the best way to check that two files are different?  I couldn't see much else in the tool that does it, except spots that were generating SHAs to be stored as fingerprints.

Reply:
this looks like it's doing a hash of the Dart objects. I imagine it would be more efficient (since we have a Uint8List anyway) to use `md5` from package:crypto: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/build_system/build_system.dart#L359. It's not cryptographically secure, but we don't care here, and it should be fast.

Reply:
(actually, I don't know if reading and hashing all the values is faster than hashing all the hash codes in the dart vm, but at least we're being consistent :) )

Reply:
Hashing is probably faster than most encryption algorithms. Also semantically we are doing a comparison here, so no reason to encrypt it (despite that it probably works too). We can either compare the hash, or loop through the bytes manually.

Reply:
you marked this resolved--are you going to use md5?

Reply:
yep.

Reply:
cool :+1: